user_profile: Add links to group settings.

tnmkr opened the pull request:
This is a possible fix for #25214
Still waiting for clarification on whether we want similar controls as in "Streams" tab or not. But i'll hazard a guess that we don't, atleast for this issue. 

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>git grep output</summary>

```bash
$ git grep .group_list_item_link

web/src/user_profile.js:    $("body").on("click", "#user-profile-modal .group_list_item_link", () => {
web/templates/user_group_list_item.hbs:            <a class = "group_list_item_link" href="{{group_edit_url}}">{{name}}</a>
```
</details>

**Screenshots and screen captures:**
<details>
<summary>Before: all users</summary>

![image](https://github.com/zulip/zulip/assets/133781250/c2dbe7ce-81b5-40c3-8167-9b8cce64e266)
</details>

<details>
<summary>After: normal users</summary>

![image](https://github.com/zulip/zulip/assets/133781250/f2c002ee-bdd7-4f5d-b31a-30f08883e6b6)

https://github.com/zulip/zulip/assets/133781250/4f8ce6e0-ffe5-4852-8714-b21562513ab5
</details>

<details>
<summary>After: guest users</summary>

![image](https://github.com/zulip/zulip/assets/133781250/c2dbe7ce-81b5-40c3-8167-9b8cce64e266)
</details>


<details>

<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks!

We should open the "All groups" tab, rather than the "Your groups", tab on the left side of the panel.

It would be good to make the same change for the stream user profile links as well (in a separate commit). Guest users should still get taken to the "Subscribed streams" tab, since that's the only one they have access to.

tnmkr commented:
### Updates

- Modified first commit to open group settings over with `All groups` tab selected when opened from user profile.
- Added another commit to open stream settings with `All streams` tab selected when opened from profile.

Screen recording of the changes made:
https://chat.zulip.org/user_uploads/2/6b/uKrse4n8QA38FArQUZaGHEAq/profile_grpsettings_dev_inteded.mp4

tnmkr commented:
### Updates

- Modified first commit to open group settings with `All groups` tab selected when opened from user profile.
- Added another commit to open stream settings with `All streams` tab selected when opened from profile.

Screen recording of the changes made:
https://chat.zulip.org/user_uploads/2/6b/uKrse4n8QA38FArQUZaGHEAq/profile_grpsettings_dev_inteded.mp4

alya commented:
@sahil839 Could you please review this one? I haven't tested the latest version.

sahil839 reviewed (commented):
Posted one comment. You should also have a look at [commit message guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages) and update the commit messages accordingy. There are some formatting and punctuation changes needed in both the commits and the second commit message description could be better by explaining the change intended more clearly without mentioning the PR number itself.

sahil839 commented:
@alya do we want to open the "All groups"/"All streams" tab when opening the settings in other places as well?

For streams - one can open the settings from stream popover, from the navbar in stream/topic narrow and by manuallly updating the hash or by opening a link pointing to the stream edit UI.

And for groups as well, the edit UI can be opened from group popover opened from group mention pill and by manuallly updating the hash or by opening a link pointing to it.

alya commented:
Good question!

1. I think if you're trying to manage groups/streams for another user, we should always open the All tab, regardless of whether you're subscribed (i.e., from someone else's user profile, though I think ideally not from your own profile).
2. If you're trying to manage your own streams/groups that you are a member of, it seems more intuitive to open the other tab. Can we check whether you are subscribed/are a member when you try to open settings for a stream/group from these other places, or does it get complicated to do that?

Might be worth flagging this question in #design.

sahil839 commented:
Sounds good. I don't think that would be tough to implement, so we can proceed with that.

tnmkr commented:
> @tnmkr can you start a discussion on CZO in #design stream.

[New CZO topic](https://chat.zulip.org/#narrow/stream/101-design/topic/stream.20and.20group.20settings.20left.20panel/near/1729835)

I'll update the PR soon.

timabbott commented:
@tnmkr I believe this is waiting for you to update to match the proposed plan in the design discussion.

tnmkr commented on web/templates/user_group_list_item.hbs:
```diff
@@ -1,5 +1,9 @@
+        {{#if is_guest}}
+            {{name}}
+        {{else}}
+            <a class = "group_list_item_link" href="{{group_edit_url}}">{{name}}</a>
```
I have removed the spaces

tnmkr commented on web/src/hashchange.js:
```diff
@@ -257,12 +264,12 @@ function do_hashchange_overlay(old_hash) {
 
         if (base === "groups") {
-            user_group_edit.change_state(section);
+            user_group_edit.change_state(section, undefined);
```
Possible bug? But from what i've tested, switching streams/groups using buttons in stream or group settings overlay doesn't trigger this. However, if navigated using links (by manually editing browser url or links in stream description), selected tab may be reset.

tnmkr replied:
For this, we can either get current toggle value and pass that OR add a possible value for `left_side_tab` like "preserve" that turns it into a no-op. The former looks cleaner, we will need to export `group_list_toggle` from `user_group_edit.js` but thats fine i think?

sahil839 replied:
I don't get what is the bug here? Can you elaborate.

tnmkr replied:
We pass `left_side_tab` as undefined in `change_state` to let it select the tab based on user's membership. So `change_state` might switch to other tab if group isn't in the currently selected tab.

For example, if I click on some link in description of a stream I am subscribed to AND that link points to an unsubscribed stream, then selected tab will change to "All streams" and list will be redrawn.

Although, I am not sure if group descriptions can contain links so might not be a problem here. But above example with streams I have tested.

tnmkr replied:
To preserve current behaviour and simplify manual testing; I've made this modification for now.

```diff
- stream_settings_ui.change_state(section, right_side_tab);
+ stream_settings_ui.change_state(section, "preserve", right_side_tab);
```

sahil839 replied:
> For example, if I click on some link in description of a stream I am subscribed to AND that link points to an unsubscribed stream, then selected tab will change to "All streams" and list will be redrawn.

Yeah, so this works fine as expected I guess. We want to show the tab based on whether the user is subscribed to the stream opened in right panel or not.

tnmkr replied:
Modified this to show appropriate tab based on membership/subscription to the group/stream being opened. Please take a look when you can @sahil839. Thanks!

tnmkr commented on web/src/user_group_edit.js:
```diff
@@ -588,11 +588,18 @@ export function change_state(section) {
+            }
+
+            // NOTE: Normally, this will update browser_history. But here we always have a group being edited, so its safe to call.
+            group_list_toggler.goto(left_side_tab);
```
Ideally i'd want to change `tabname` inside `redraw_user_group_list` to avoid rewriting the previously drawn list here. However, i need to call `group_list_toggler.goto()` anyway to change the active tab in widget (`redraw_left_panel` in `redraw_user_group_list` simply populates the list, selected tab is managed by `group_list_toggler`). So i've put this to make it clear whats going on.

sahil839 replied:
Can you just remove the `redraw_user_group_list()` call above as I guess that would be called when calling `goto`.

tnmkr replied:
I have removed the call from here. `stream_settings_ui` has some comments stating redraws being needed for legacy reasons. So still calling `goto` after `redraw_left_panel` there.

tnmkr reviewed (commented):
All requested features have been implemented. But i am not sure if the code changes i've made are good. 

Note that i've put "comments" on either group_setting or stream_settings here. But i need guidance on both, since they are almost symmetrical.

tnmkr commented on web/src/stream_settings_ui.js:
```diff
@@ -750,7 +760,7 @@ export function change_state(section, right_side_tab) {
 }
 
-export function launch(section, right_side_tab) {
+export function launch(section, left_side_tab, right_side_tab) {
```
Probably documenting this behaviour near definition of `change_state` and a blueslip warning will suffice?

tnmkr commented:
@sahil839 I have updated this PR and have some questions regarding implementation above. Could you please take a look at this when you are free? Thanks!

sahil839 commented on web/src/hashchange.js:
```diff
@@ -58,6 +58,13 @@ function show_all_message_view() {
+function is_somebody_else_profile_open() {
+    return (
+        user_profile.get_user_id_if_user_profile_modal_open() !== undefined &&
+        !people.is_my_user_id(user_profile.get_user_id_if_user_profile_modal_open())
```
Can we instead do `user_profile.get_user_id_if_user_profile_modal_open() === my_current_user_id` here?

tnmkr replied:
Like this? I am doing that first check to make sure a profile is open.
```js
    return (
        user_profile.get_user_id_if_user_profile_modal_open() !== undefined &&
        user_profile.get_user_id_if_user_profile_modal_open() !== people.my_current_user_id()
    );
```

sahil839 replied:
So, `people.my_current_user_id` will not be undefined in any case except spectators and we anyways do not allow spectators to access stream or groups url. So we can directly compare it like described in the above comment, but we may use this function in future for spectators in some case and so we can keep it as it is for now.

sahil839 replied:
I got confused, the latest change in the PR is fine, no need to do any more changes for this.

sahil839 commented:
Posted one comment and replies to your comments. I have a look at the approach and looks good at a first glance. Would review it again once you rebase the PR since the url handling code has changed recently.

tnmkr commented:
@sahil839 I have rebased the PR on main. Also made some changes to possibly address concerns raised in comments. Please take a look when you can. Thanks!

sahil839 commented on web/src/user_group_edit.js:
```diff
@@ -573,11 +573,26 @@ export function change_state(section, right_side_tab) {
+        if (left_side_tab === "preserve") {
+            redraw_user_group_list();
+            switch_to_group_row(group);
+            return;
```
Do we use this code?

tnmkr commented on web/src/user_group_edit.js:
```diff
@@ -573,11 +573,26 @@ export function change_state(section, right_side_tab) {
+        // group is being edited. We are always editing a group here
+        // so its safe to call
+        set_active_group_id(group.id);
+        group_list_toggler.goto(left_side_tab);
```
> Can we just call `switch_to_group_row` first and then call `goto`, this way we can avoid calling `set_active_group_id` here?

I had it like that, however while testing i found that the selected group was not visible(without scrolling) if it was at the bottom of the list. Calling `switch_to_group_row` after `goto`, scrolls the group into view. I feel like this makes for better ux as the user can tell why we are selecting this tab by default.

> Also, can we add a condition to call `goto` only if the current selected tab is not equal to `left_side_tab`, both here and in stream settings code?

Good idea! I don't think that'll cause any issues. Will check and do.

tnmkr commented:
Updates in latest push:
- Fixed typo in commit message
- Made it so `left_side_tab` isn't actually switched to if the new tab is already selected. [Comment](https://github.com/zulip/zulip/pull/28654#discussion_r1508828041)
- Removed unused code regarding `left_side_tab === preserve`. [Comment](https://github.com/zulip/zulip/pull/28654#discussion_r1508826212)

sahil839 commented:
Looks good to me.

@timabbott this is ready for your review.

timabbott commented on web/src/hashchange.js:
```diff
@@ -345,7 +352,17 @@ function do_hashchange_overlay(old_hash) {
+        if (is_somebody_else_profile_open()) {
+            user_group_edit.launch(section, "all-groups", right_side_tab);
+            return;
+        }
```
It feels somewhat unclean to me to be doing this profile check here; I think it would be more cleanly done by registering a click handler for these group elements rather than relying solely on the URL, and using that to pass through a parameter. But doing that might be a pain, so I'm not sure it's worth blocking this PR on that.
